config-maps-for-spark-confs.md

```diff
@@ -0,0 +1,149 @@
+# ConfigMaps for Spark Configuration Files
+
+## Introduction
+
+This specification outlines methods for exposing Spark configuration
+files through ConfigMaps.
+
+## Problem statement
+
+Currently the only way to customize spark configuration files such
+as log4j.properties and spark-defaults.conf for clusters
+launched from oshinko is to create a custom spark image and redeploy
+oshinko to use that image. This is a very heavy-weight approach for
+configuration, especially in situations where cluster or application
+tuning are being done. Oshinko needs a more dynamic way to control
+the cluster configuration files.
+
+## Proposed solution
+
+ConfigMap objects mounted as volumes on the spark master and worker
+nodes would provide a way to modify the default configuration for
+a cluster without creating a new image.
+
+ConfigMap objects can store multiple named configuration files and
+be mounted to a particluar directory in a container. Consquently, it
+is possible to specify log4j.properties, spark-defaults.conf, and other
+files in a ConfigMap and mount them at a location such as /opt/spark/oshinko-conf
+in a spark node.  If the environment variable SPARK_CONF_DIR is set
+to /opt/spark/oshinko-conf, then spark will read its configuration from
+the ConfigMap.
+
+There are two approaches to using ConfigMaps for spark configuration
+in oshinko.  The simpler approach will be outlined here, and the more
+complex approach briefly described under `Alternatives`. Since this is
+an important feature related to cluster/application tuning and performance
+metrics, this spec recommends pursuing the simple approach first which can
+be done in a more timely manner and then perhaps implementing the alternative
+at a later time.
+
+In the simple approach, a `sparkConfig` field is added to the cluster
+config JSON object which names a ConfigMap to use as a source for spark
+cluster configuration:
+
+    {
+      "name": "fred",
+      "config":
+              {
+                "masterCount": 1,
+                "workerCount": 2,
+                "sparkConfig": "myconfiguration"
```

Comment:
That sounds workable, and the nice thing is that you would know by looking at the config for the cluster that you overwrote particular files, whereas with sparkConfig you only have a single configmap but have no idea what you overwrote.

If we verify that there is a simple way to make these from files using "oc", I think this is a great suggestion.